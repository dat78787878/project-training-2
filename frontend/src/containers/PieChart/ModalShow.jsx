import React from 'react';
import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getData } from '../../redux/pieChart/actions';
import { Modal, Button } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck } from '@fortawesome/free-solid-svg-icons';
import { difference } from 'lodash';
const ModalShow = () => {
  const pieData = useSelector((state) => state.pieChart.pieData);
  const { fromDate, toDate } = useSelector((state) => state.dateRange);
  const listAllData = ['Android', 'Windows', 'iOS', 'OsX', 'Unknown', 'Linux'];
  const [listDataModal, setListDataModal] = useState(listAllData);
  const [listDataCheck, setListDataCheck] = useState(pieData.map((val) => val.x));
  const [show, setShow] = useState(false);

  const dispatch = useDispatch();

  const onClickLabel = (val) => {
    setListDataCheck((prevList) => {
      const isChecked = prevList.includes(val);
      if (isChecked) {
        return prevList.filter((check) => check !== val);
      }
      return [...prevList, val];
    });
  };

  const onHide = () => {
    const prevListDataCheck = pieData.map((val) => val.x);
    difference(listDataCheck, prevListDataCheck)
      ? dispatch(getData([fromDate, toDate, listDataCheck]))
      : '';
    setShow(false);
  };

  const handleSearch = (e) => {
    e.preventDefault();
    const keyFilter = e.target.value.trim().toLowerCase();
    if (keyFilter.length > 0) {
      setListDataModal((prev) => {
        return prev.filter((val) => val.toLowerCase().includes(keyFilter));
      });
    } else {
      setListDataModal(listAllData);
    }
  };

  return (
    <div className="modalShow">
      <Button variant="success" className="m-1" onClick={() => setShow(true)}>
        Show modal
      </Button>

      <Modal
        show={show}
        onHide={onHide}
        dialogClassName="modal-custom"
        aria-labelledby="example-custom-modal-styling-title">
        <Modal.Body>
          <input className="modal-custom-input" onChange={handleSearch} />
          {listDataModal.map((val, index) => {
            return (
              <div
                key={index}
                className="d-flex justify-content-between p-2"
                onClick={() => onClickLabel(val)}>
                <span>{val}</span>
                {listDataCheck.includes(val) && <FontAwesomeIcon icon={faCheck} />}
              </div>
            );
          })}
        </Modal.Body>
      </Modal>
    </div>
  );
};

export default ModalShow;
