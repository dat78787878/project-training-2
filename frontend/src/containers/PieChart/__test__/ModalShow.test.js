import React from 'react';
import ModalShow from '../ModalShow';
import { useSelector, useDispatch } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

describe('Modal', () => {
  const mockState = {
    pieChart: {
      pieData: [
        { x: 'Android', y: 20 },
        { x: 'Windows', y: 30 },
        { x: 'iOS', y: 60 },
        { x: 'Os X', y: 40 },
        { x: 'Unknown', y: 10 },
        { x: 'Linux', y: 50 }
      ],
      isLoading: false,
      isError: false
    },
    dateRange: {
      fromDate: '',
      toDate: ''
    }
  };
  const mockDispatch = jest.fn();
  beforeEach(() => {
    useDispatch.mockImplementation(() => mockDispatch);
    useSelector.mockImplementation((callback) => {
      return callback(mockState);
    });
  });
  afterEach(() => {
    useDispatch.mockClear();
    useSelector.mockClear();
  });

  it('render modal', () => {
    const { container } = render(<ModalShow />);
    expect(container.getElementsByClassName('modalShow'));
    expect(container).toMatchSnapshot();
  });
  it('Filter input', () => {
    const { queryByText, getByRole } = render(<ModalShow />);
    const buttonLabel = getByRole('button', { label: /button-name/i });
    fireEvent.click(buttonLabel);
    const inputFilter = getByRole('textbox');
    fireEvent.change(inputFilter, { target: { value: 'android' } });
    expect(inputFilter.value).toEqual('android');
    expect(queryByText('iOS')).not.toBeTruthy();
    fireEvent.change(inputFilter, { target: { value: '' } });
    expect(queryByText('iOS')).toBeTruthy();
  });
  it('Click label', () => {
    const { getByText, getByRole, queryAllByTestId } = render(<ModalShow />);
    const buttonLabel = getByRole('button', { label: /button-name/i });
    fireEvent.click(buttonLabel);
    const label = getByText('Android');
    fireEvent.click(label);
    expect(queryAllByTestId('check')).toHaveLength(5);
    fireEvent.click(label);
    expect(queryAllByTestId('check')).toHaveLength(6);
  });
});
