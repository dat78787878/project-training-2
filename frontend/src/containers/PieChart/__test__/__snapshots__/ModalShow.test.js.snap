// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal render modal 1`] = `
<div>
  <div
    class="modalShow"
  >
    <button
      aria-label="button-name"
      class="m-1 btn btn-success"
      type="button"
    >
      Show modal
    </button>
  </div>
</div>
`;
